gufe/protocols/protocolunit.py

```diff
@@ -7,79 +7,312 @@
 """
 
 import abc
+from dataclasses import dataclass
+import traceback
+import uuid
 from os import PathLike
+from copy import copy
 from typing import Iterable, List, Dict, Any, Optional, Union
+import tempfile
 
-from .results import (
-    ProtocolUnitResult, ProtocolUnitFailure,
-)
+from ..tokenization import GufeTokenizable, GufeKey, normalize
 
 
-class ProtocolUnit(abc.ABC):
-    """A unit of work within a `ProtocolDAG`"""
+@dataclass
+class Context:
+    """Data class for passing around execution context components to
+    `ProtocolUnit._execute`.
+
+    """
+    unit_scratch: PathLike
+    dag_scratch: PathLike
+
+
+class ProtocolUnitResultBase(GufeTokenizable):
+    inputs: Dict[str, Any]   # `ProtocolUnitResult`s this `ProtocolUnitResult` is dependent on
+
+    name: Optional[str]      # name of the `ProtocolUnit` that produced this `ProtocolUnitResult`
+    pure: bool               # whether `ProtocolUnit` that produced this `ProtocolUnitResult` was a function purely of its inputs
+    source_key: GufeKey             # key of the `ProtocolUnit` that produced this `ProtocolUnitResult`
+
+    outputs: Dict[str, Any]  # outputs is a dict returned by a `ProtocolUnit`'s `_execute` method
+
+    def __init__(self, *, 
+            name: Optional[str] = None, 
+            source_key: GufeKey, 
+            pure: bool, 
+            inputs: Dict[str, Any], 
+            outputs: Dict[str, Any]
+        ):
+        """Generate a `ProtocolUnitResult`.
+
+        Parameters
+        ----------
+        name : Optional[str]
+            Name of the `ProtocolUnit` that produced this `ProtocolUnitResult`.
+        source_key : GufeKey
+            Key of the `ProtocolUnit` that produced this `ProtocolUnitResult`
+        pure : bool
```

Comment:
Why have pure? Isn't uuid fine at this level? One less branch to think about

Reply:
You're right that making `ProtocolUnit`s hash based on their inputs isn't strictly necessary, but does offer opportunities for deduplication within a Scheduler. So, it's an optimization.

The way I have it now, `pure` is default `False` (taking cues from `dask.delayed` behavior), so it's opt-in. My view is that if protocol developers don't generally use this feature, we can remove it.

Reply:
I went ahead and removed `pure` as a concept, and switched to just using UUIDs for `ProtocolUnit`s. I think the opportunities for deduplication across DAGs may prove rather limited, so this simplifies things here.